--- /Users/ethan/Src/github.com/scrcpy-mobile/porting/../scrcpy/server/src/main/java/com/genymobile/scrcpy/FakeContext.java	2025-12-31 18:59:43
+++ /Users/ethan/Src/github.com/scrcpy-mobile/porting/server/src/main/java/com/genymobile/scrcpy/FakeContext.java	2025-12-31 19:05:22
@@ -16,6 +16,7 @@
 
 public final class FakeContext extends ContextWrapper {
 
+    // Fix android 16 beta audio issue: https://github.com/Genymobile/scrcpy/issues/5698#issuecomment-2559277928
     public static final String PACKAGE_NAME = "com.android.shell";
     public static final int ROOT_UID = 0; // Like android.os.Process.ROOT_UID, but before API 29
 
